src/state/slices/opportunitiesSlice/resolvers/osmosis/utils.ts

```diff
@@ -0,0 +1,198 @@
+import axios from 'axios'
+import { getConfig } from 'config'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { logger } from 'lib/logger'
+
+type OsmosisToken = {
+  denom: string
+  amount: string
+}
+
+type OsmosisPoolAsset = {
+  token: OsmosisToken
+  weight: string
+}
+
+type OsmosisPool = {
+  '@type': string
+  name: string
+  address: string
+  id: string
+  pool_params: {
+    swap_fee: string
+    exit_fee: string
+    smooth_weight_change_params: boolean
+  }
+  future_pool_governor: string
+  total_shares: {
+    denom: string
+    amount: string
+  }
+  pool_assets: OsmosisPoolAsset[]
+  total_weight: string
+  apy: string
+  tvl: string
+}
+
+type OsmosisBasePool = Omit<OsmosisPool, 'apy' | 'tvl' | 'underlyingAssets'>
+
+type OsmosisPoolList = {
+  pools: OsmosisBasePool[]
+  pagination: {
+    next_key: string | null
+    total: string
+  }
+}
+
+type PoolHistoricalData = {
+  symbol: string
+  amount: number
+  denom: string
+  coingecko_id: string
+  liquidity: number
+  liquidity_24h_change: number
+  volume_24h: number
+  volume_24h_change: number
+  volume_7d: number
+  price: number
+  fees: string
+  main: boolean
+}
+
+type PoolHistoricalDataList = {
+  [key: string]: PoolHistoricalData[]
+}
+
+const moduleLogger = logger.child({
+  namespace: ['opportunitySlice', 'resolvers', 'osmosis', 'utils'],
+})
+
+/** Somehow, the v1beta1/pools API call returns some (~12/886) objects of a type that doesn't conform
+ * to the type defined above (missing certain fields and containing others.)
+ * This type guard is used to filter out the invalid pools records. */
+const isOsmosisBasePool = (pool: any): pool is OsmosisBasePool => {
+  return (
+    pool['@type'] !== undefined &&
+    pool.address !== undefined &&
+    pool.id !== undefined &&
+    pool.pool_params !== undefined &&
+    pool.pool_liquidity === undefined &&
+    pool.future_pool_governor !== undefined &&
+    pool.total_shares !== undefined &&
+    pool.pool_assets !== undefined &&
+    pool.total_weight !== undefined
+  )
+}
+
+export const generateAssetIdFromOsmosisDenom = (denom: string): AssetId => {
+  if (denom.startsWith('u') && denom !== 'uosmo') {
+    return toAssetId({
+      assetNamespace: 'native' as const,
+      assetReference: denom,
+      chainId: osmosisChainId,
+    })
+  }
+
+  if (denom.startsWith('ibc')) {
+    return toAssetId({
+      assetNamespace: 'ibc' as const,
+      assetReference: denom.split('/')[1],
+      chainId: osmosisChainId,
+    })
+  }
+
+  return toAssetId({
+    assetNamespace: 'slip44' as const,
+    assetReference: ASSET_REFERENCE.Osmosis,
+    chainId: osmosisChainId,
+  })
+}
+
+export const getPools = async (): Promise<OsmosisPool[]> => {
+  try {
+    /**
+     * TODO: Use axios cache layer with reasonable (30s-5m) max age to save responses between calls.
+     * At app startup, these requests are made ~10 times with the same data returned each time.
+     */
```

Comment:
rather than this - pop this in an RTK query itself, and you can get the result of it inside other resolvers

Reply:
or another option is just lodash `memoize`

Reply:
If we fix the chainId early return issue, this will be fine since the consumer resolver will not be executing more than it needs to - RTK is probably overkill for this but `memoize` is a low-cost safety we can wear just in case IMO

Reply:
@0xdef1cafe @gomesalexandre 

I'm not sure if this is a corner case worth considering, but what about the event where memoized network request fails? `_.memoize()` has no mechanism to handle this and will cache the result of the failed request. Likewise, if the promise rejects, `_.memoize()` will cache the failed promise and will not run the unmemoized function again.